Library/Homebrew/utils/io_selector.rb

```diff
@@ -0,0 +1,67 @@
+require "delegate"
+
+require "extend/io"
+
+module Utils
+  class IOSelector < DelegateClass(Hash)
+    attr_reader :separator
+
+    alias all_streams keys
+    alias all_tags values
+    alias tag_of fetch
+
+    def self.each_line_from(streams = {}, separator = $/, &b)
+      new(streams, separator).each_line_nonblock(&b)
+    end
+
+    def each_line_nonblock(&b)
+      each_line_nonblock_enum.each(&b)
+    end
+
+    def pending_streams
+      @pending_streams ||= all_streams.dup
+    end
+
+    def initialize(streams = {}, separator = $/)
+      super(streams.invert.compare_by_identity)
+      @separator = separator
+    end
+
+    private
+
+    def each_line_nonblock_enum
+      Enumerator.new do |yielder|
```

Comment:
Oh, this is in fact one of my favorite concepts in the Ruby standard library! :blush:

The benefit here is that we can invoke `each_line_nonblock` _without a block_ – as a second form of calling the method, so to speak.

Not having a block causes `b == nil`, so we’re basically calling `#each(nil)` on the enumerator, which (by contract) means that the enumerator returns itself.

So the wrapper gives us a second calling form for free:

1. **each_line_nonblock { |tag, line| } → obj**

2. **each_line_nonblock → Enumerator**

This, in turn, allows client code to do useful and concise things like e. g.:

```
selector.each_line_nonblock.map { |_, line| line }.join("")
```

> instead of simply calling this directly […]?

Inlining this would mean we’d have to go with `do […] end.each`, which would go against our style guide, and Rubocop would certainly want to have a word with us. :wink: